xds/src/main/java/io/grpc/xds/internal/Matchers.java

```diff
@@ -0,0 +1,328 @@
+    }
+
+    /** The input string should have the prefix. */
+    public static StringMatcher forPrefix(String prefix, boolean ignoreCase) {
+      checkNotNull(prefix, "prefix");
+      return StringMatcher.create(null, prefix, null, null, null,
+          ignoreCase);
+    }
+
+    /** The input string should have the suffix. */
+    public static StringMatcher forSuffix(String suffix, boolean ignoreCase) {
+      checkNotNull(suffix, "suffix");
+      return StringMatcher.create(null, null, suffix, null, null,
+          ignoreCase);
+    }
+
+    /** The input string should match this pattern. */
+    public static StringMatcher forSafeRegEx(Pattern regEx) {
+      checkNotNull(regEx, "regEx");
+      return StringMatcher.create(null, null, null, regEx, null,
+          false/* doesn't matter */);
+    }
+
+    /** The input string should contain this substring. */
+    public static StringMatcher forContains(String contains) {
+      checkNotNull(contains, "contains");
+      return StringMatcher.create(null, null, null, null, contains,
+          false/* doesn't matter */);
+    }
+
+    /** Returns the matching result for this string. */
+    public boolean matches(String args) {
+      if (args == null) {
+        return false;
+      }
+      if (exact() != null) {
+        return ignoreCase()
+            ? exact().equalsIgnoreCase(args)
+            : exact().equals(args);
+      } else if (prefix() != null) {
+        return ignoreCase()
+            ? args.toLowerCase().startsWith(prefix().toLowerCase())
+            : args.startsWith(prefix());
+      } else if (suffix() != null) {
+        return ignoreCase()
+            ? args.toLowerCase().endsWith(suffix().toLowerCase())
+            : args.endsWith(suffix());
+      } else if (contains() != null) {
+        return args.contains(contains());
+      }
+      return regEx().matches(args);
+    }
+
+    private static StringMatcher create(@Nullable String exact, @Nullable String prefix,
+        @Nullable String suffix, @Nullable Pattern regEx, @Nullable String contains,
+        boolean ignoreCase) {
+      return new AutoValue_Matchers_StringMatcher(exact, prefix, suffix, regEx, contains,
+          ignoreCase);
+    }
+  }
+
+  /** Matcher to evaluate whether an IPv4 or IPv6 address is within a CIDR range. */
+  @AutoValue
+  public abstract static class CidrMatcher {
+
+    abstract InetAddress addressPrefix();
+
+    abstract int prefixLen();
+
+    /** Returns matching result for this address. */
+    public boolean matches(InetAddress address) {
+      if (address == null) {
+        return false;
+      }
+      byte[] cidr = addressPrefix().getAddress();
+      BigInteger cidrInt = new BigInteger(cidr);
+      BigInteger addrInt = new BigInteger(address.getAddress());
+
+      int shiftAmount = 8 * cidr.length - prefixLen();
+
+      cidrInt = cidrInt.shiftRight(shiftAmount);
+      addrInt = addrInt.shiftRight(shiftAmount);
+      return cidrInt.equals(addrInt);
```

Comment:
`0.0.0.1` should not match `::1`. This code needs to check the IP version or length as well.